### ISSUE TITLE
LSTM for Sequence Classification into Categories

### ISSUE THREAD

**Issue by brunoalano** [fchollet/keras#2227]
Hi,

I'm trying to use a LSTM where I provide `word2vec` vectors as inputs and I want to classify the this sequence into a set of categories, so, as training data, I have something like that for example:

```
"<text>", <category id>
"lorem ipsum dolor sit amet...", 5
```

So, my question is: how can I feed that into my LSTM?

Here is how I build my model:

``` python
# Create the LSTM Neural Model
model = Sequential()
model.add(
  Embedding(
    output_dim=word_vector_model.dimension,
    input_dim=number_of_words_in_dataset, # thats correct?
    mask_zero=True,
    weights=[embedding_weights], # what this should be?
    input_length=input_length, # what this should be?
    output_dim=number_of_categories # thats correct?
  )
)
model.add(LSTM(word_vector_model.dimension))
model.add(Dropout(0.5)) # try 0.3
model.add(Dense(1, activation='sigmoid'))

# Compile the model
model.compile(loss='binary_crossentropy', optimizer='adam')
```
- What `output_dim` should be? The number of categories in `y`?
- What are the `embedding_weights`?
- What `input_dim` should be?

Thank you, and sorry if I don't understand the meaning of each field.

**Comment by braingineer** [comment]
I personally use a colleague's vocabulary class (you can find his implementation [here](https://github.com/timvieira/arsenal/blob/master/arsenal/alphabet.py) .  Basically, it maps string -> int.

You can do (and I prefer the nomenclature vocab... sorry if that's confusing):

```python
## assuming dataset is an array of arrays of words. 
vocab = Alphabet.from_iterable(word for sent in dataset for word in sent)
## if not.. 
vocab  = Alphabet.from_iterable(word for sent in dataset for word in sent.split(" "))
```

Now you have indices you can use.    Some other things to consider:

1. if you want to use a mask on the lstm for variable lengths, you'll not want the vocab to map a word to the 0th index
2. You want the indices of the vocabulary to match the indices on your embedding matrix so that the embedding layer correctly works. I preferably make a new matrix by doing (for example, and using glove which comes in a text file):

```python:
def make_embeddings(igor, vocab):
    #glove_fp = "/research/data/glove/glove.840B.300d.txt"
    vocab_togo = set(vocab._mapping.keys())
    E = igor.embedding_size
    V = len(vocab)
    embeddings = np.zeros((V, E))
    with open(igor.glove_fp) as fp:
        for line in tqdm(fp.readlines(), unit='lines'):
            line = line.replace("\n","").split(" ")
            word,nums = line[0], [float(x.strip()) for x in line[1:]]
            if word in vocab_togo:
                vocab_togo.remove(word)
                embeddings[vocab[word]] = np.array(nums)

    print("{} words not in glove".format(len(vocab_togo)))
    with open("glove_oov.txt", "w") as fp:
        fp.write("\n".join(vocab_togo))

    for word in vocab_togo:
        embeddings[vocab[word]] = GlorotUniform().sample((E,1)).reshape(E)

    utils.save_embeddings(igor, embeddings, vocab)
    return embeddings
```

**Comment by AAMIBhavya** [comment]
@brunoalano , I am also working with the same task. Did you solve the problem?

**Comment by vinayakumarr** [comment]
Any body having character based text classification code in keras like in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/skflow/text_classification_character_rnn.py
